org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/controlflow/ConstantConditionsInterpreter.xtend

```diff
@@ -355,6 +355,13 @@ class ConstantConditionsInterpreter {
 	}
 	
 	def dispatch EvaluationResult internalEvaluate(XCastedExpression expression, EvaluationContext context) {
+		val declaredType = context.resolvedTypes.getActualType(expression.target)
+		if (declaredType?.primitive) {
+			val castedType = context.resolvedTypes.getActualType(expression)
+			if (castedType?.identifier != declaredType?.identifier) {
```

Comment:
DeclaredType is definitly not null here (see null-check above).
It's safe to use castedType?.type != declaredType.type.

My problem with this solution is, that JDT reports `if ((byte) 1 == (short) 1) ..` as a constant condition. Do we need to check value ranges and floating / integral number casts, instead?